File: problems/0238-product-of-array-except-self/solution.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {

        int zcount = 0;
        int tprod = 1;
        int tprod_exclude_zeros = 1;
         vector<int> sol;

        // Loop once to get zero count and product totals
        for (int n : nums) {
            if (n == 0) {
                zcount++;
            } else {
                tprod_exclude_zeros *= n;
            }
            tprod *= n;
        }

        // loop once more to create solution array
        if (zcount == 0) {
            for (int n : nums) {
                sol.push_back(tprod / n);
            }
        } else if (zcount == 1) {
            for (int n : nums) {
                if (n != 0) {
                    sol.push_back(0);
                } else {
                    sol.push_back(tprod_exclude_zeros);
                }
            }
        } else if (zcount > 1) {
            for (int n : nums) {
                sol.push_back(0);
            }
        }

        return sol;

    }
};

File: problems/0238-product-of-array-except-self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ltr, rtl = [], []


        prod = 1
        for n in nums:
            prod *= n
            ltr.append(prod)
        

        prod = 1
        for i in range(len(nums) - 1, -1, -1):
            prod *= nums[i]
            rtl.append(prod)
        rtl.reverse()


        ans = []
        for i in range(len(nums)):
            if i == 0:
                ans.append(rtl[1])
            elif i == len(nums) - 1:
                ans.append(ltr[-2])
            else:
                ans.append(ltr[i - 1] * rtl[i + 1])
        
        return ans

